### ISSUE TITLE
Default train batch size set to 1

### ISSUE THREAD

**Issue by ksteimel** [mozilla/DeepSpeech#2204]
- **Have I written custom code (as opposed to running examples on an unmodified clone of the repository)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04
- **TensorFlow installed from (our builds, or upstream TensorFlow)**: Upstream tensorflow
- **TensorFlow version (use command below)**: 1.13.1 (v1.13.1-0-g6612da8951)
- **Python version**:  3.6.8
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 10.1/
- **GPU model and memory**: Gigabyte GTX 1070 (8gb)
- **Exact command to reproduce**: python3 DeepSpeech.py --train_files <train_file.csv> --dev_files <dev_file.csv> --test_files <test_file.csv>

If the `--train_batch_size` flag is not added for training, the default batch size is 1. This is a bug because the model will not train as well if the weights are updated this frequently. This also causes very inefficient use of gpus as most time is spent shuffling data across the PCIe connector and not training.

**Comment by reuben** [comment]
There's no better default we can use. A higher value will cause OOMs on people with low RAM. Training a model is not supposed to be a drag and drop thing, you're expected to tune the hyperparameters as appropriate for your data/system/use case.

**Comment by ftyers** [comment]
@reuben this tripped us up today. Perhaps there are some other options:

1. print a warning if people have a batch size of one? 
2. make `--train_batch_size` a mandatory argument ?  -- if people are expected to change it, maybe it shouldn't have default ? 

I agree that expecting people to tune the hyperparameters is reasonable, but I think that there is possibly a way to save people time by anticipating "typical error cases".

**Comment by reuben** [comment]
Removing the default seems reasonable.

**Comment by reuben** [comment]
https://github.com/mozilla/DeepSpeech/pull/2205

**Comment by reuben** [comment]
So #2205 turned out not to be a working approach as we have execution flows through DeepSpeech.py where the batch size flags are optional. If anyone has a better idea on how to fix this in a clean way, feel free to propose it.

**Comment by dabinat** [comment]
Seeing as the batch size is dependent upon VRAM, perhaps DeepSpeech could detect the VRAM and automatically choose an appropriate value?

**Comment by kdavis-mozilla** [comment]
@dabinat This is not easy to do.

Someone may change the width of the network which would change the maximum batch size. Google may change the implementation of LSTM's which would change the maximum batch size. NVIDIA may change the implementation of LSTM's which would change the maximum batch size.... There are many other similar problematic scenarios.

It seems that this cure may be worse than the sickness.

**Comment by reuben** [comment]
Closing as not actionable.
